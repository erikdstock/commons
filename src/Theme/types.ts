import { SystemStyleObject, Theme, ThemeValue } from "@styled-system/css"
import * as CSS from "csstype"

export type ThemeColors = ThemeValue<CSS.ColorProperty>
export type ThemeFonts = ThemeValue<CSS.FontFamilyProperty>

type ObjectOrArray<T> = T[] | { [K: string]: T | ObjectOrArray<T> }

// interface BaseThemeKnownColors {
//   [k: string]: CSS.ColorProperty
//   // text: CSS.ColorProperty
//   // background: CSS.ColorProperty
//   // primary: CSS.ColorProperty
//   // secondary: CSS.ColorProperty
//   // muted: CSS.ColorProperty
// }

interface Variants {
  variants?: {
    [k: string]: SystemStyleObject
  }
}

type SystemStyleWithVariants = SystemStyleObject & Variants

// interface RosesThemeColors extends BaseThemeKnownColors {
//   [k: string]: CSS.ColorProperty | ObjectOrArray<CSS.ColorProperty>
// }

/** The roses theme follows the styled-system/system-ui theme specification */
export type RosesThemeObject = Theme & {
  // colors: RosesThemeColors
  /** Theme-ui html styles (incoming from baseTheme) */
  styles?: { [elementName: string]: SystemStyleWithVariants }
  /** Roses html styles */
  htmlStyles?: { [elementName: string]: SystemStyleObject }
  /** Roses component styles */
  componentStyles?: { [componentName: string]: SystemStyleWithVariants }
  /** Custom component variant styles */
}
