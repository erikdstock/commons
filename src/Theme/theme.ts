import { Theme } from "styled-system"

const colors = {
  red: "#db1d33",
  black: "#000",
  white: "#fff",
  gray: {
    "1": "hsl(0,0%,32%)",
    "2": "hsl(0,0%,47%)",
    "3": "hsl(0,0%,58%)",
    "4": "hsl(0,0%,68%)",
    "5": "hsl(0,0%,77%)",
    "6": "hsl(0,0%,85%)",
    "7": "hsl(0,0%,93%)",
  },
}

// Default Breakpoints
const breakpoints = ["40em", "52em", "64em"]

// default fontSizes
const fontSizes = [12, 14, 16, 20, 24, 32, 48, 64, 72]

// default space for margin and padding
const space = [0, 4, 8, 16, 32, 64, 128, 256, 512]

// type NestedObjectOrArray<T> = T[] | { [K: string]: T | NestedObjectOrArray<T> }

interface CommonsTheme extends Theme {
  // colors: NestedObjectOrArray<CSS.ColorProperty>
  colors: any
}

export const defaultTheme: CommonsTheme = {
  breakpoints,
  colors,
  fontSizes,
  space,

  fonts: {
    serif: "'IBM Plex Serif', serif",
    sans: "Chivo, Helvetica, Sans",
    mono: "'Fira Mono', monospace",
  },

  radii: [0, 2, 3, 5, 10],
  buttons: {
    primary: {
      color: colors.white,
      backgroundColor: colors.red,
    },
    outline: {
      color: colors.red,
      backgroundColor: colors.white,
      boxShadow: "inset 0 0 0 2px",
    },
    inactive: {
      backgroundColor: colors.gray[5],
      color: colors.white,
      disabled: true,
    },
  },
}
