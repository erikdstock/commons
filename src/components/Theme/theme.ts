import { NONAME } from "dns";

/**
 * An interface matching what what rebass expects.
 * See https://rebassjs.org/theming
 */
export interface RebassTheme {
  /* Array of strings representing viewport widths to use for min-width media queries. */
  breakpoints?: string[]
  
  /* Array of numbers to use as a typographic scale */
  fontSizes?:	[number|string]
  
  /* Color names that can be used in color, bg, and borderColor props (nested objects can be accessed via dot notation) */
  colors?: object
  
  /* Array of numbers for use as margin and pixel values */
  space?: object

  /* Values for the fontFamily prop */
  fonts?: object
  
  /* Array or Object values for fontWeight prop */
  fontWeights?: object

  /* Array or Object	Values for lineHeight prop */
  lineHeights?: object

  /* 	Array or Object	Values for letterSpacing prop */
  letterSpacings?: object

  /* Array or Object	Values for boxShadow prop */
  shadows?:	object

  /* Array or Object	Values for border props */
  borders?: object

  /* Array or Object	Values for borderRadius props */
  radii?: object

  /* Array or Object	Values for opacity props */
  opacity?: object

  /* Styles for button varians */
  buttons: object
}

const colors = {
  red: "#db1d33",
  black: "#000",
  white: "#fff",
  gray: {
    "1": "hsl(0,0%,32%)",
    "2": "hsl(0,0%,47%)",
    "3": "hsl(0,0%,58%)",
    "4": "hsl(0,0%,68%)",
    "5": "hsl(0,0%,77%)",
    "6": "hsl(0,0%,85%)",
    "7": "hsl(0,0%,93%)",
  }
}

export const defaultTheme: RebassTheme = {
  fonts: {
    serif: "'IBM Plex Serif', serif",
    sans: "Chivo, Helvetica, Sans",
    mono: "'Fira Mono', monospace"
  },

  colors,
  buttons: {
    primary: {
      color: colors.white,
      backgroundColor: colors.red,
    },
    outline: {
      color: colors.red,
      backgroundColor: colors.white,
      boxShadow: 'inset 0 0 0 2px'
    },
    inactive: {
      backgroundColor: colors.gray[5],
      color: colors.white,
      disabled: true
    },
  }
}