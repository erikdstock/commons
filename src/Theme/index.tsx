import React from "react";
import { ThemeProvider } from "styled-components";
import { defaultTheme, RebassTheme } from "./theme";

export { defaultTheme, RebassTheme };

interface Props {
  theme?: RebassTheme;
}


/**
 * A theme provider
 */
export const Theme: React.FunctionComponent<Props> = ({
  children,
  theme = defaultTheme
}) => (
  <ThemeProvider theme={theme}>
    <>
      {
        /* ThemeProvider Children must be a node, so Fragment </> is required */
        children
      }
    </>
  </ThemeProvider>
);
