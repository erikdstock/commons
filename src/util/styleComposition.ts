import styled from "@emotion/styled"
import { CSSObject, SystemStyleObject } from "@styled-system/css"
import ssCss from "@styled-system/css"
import get from "lodash/get"
import { RosesTheme } from "../theme"

/** Clumsily force styledCss function to believe that there will be a theme present */
type StyledCssFn = (o: SystemStyleObject) => CSSObject
const styledCss = ssCss as StyledCssFn

/** The base style for Box and everything on up. */
export const boxStyle: CSSObject = {
  boxSizing: "border-box",
}

interface ComposeOptions {
  defaultVariant?: string
}

const themeVariantsRoot: keyof RosesThemeObject = "variants"
const componentStylesRoot: keyof RosesThemeObject = "componentStyles"

interface StyledInterpolationProps extends RosesStyleProps {
  theme: RosesThemeObject
}

type InterpolationFn = (p: StyledInterpolationProps) => CSSObject | void

/** Reach into the theme, get a style object, apply it with styledCss */
const themedComponent: (
  path: string
) => (p: StyledInterpolationProps) => CSSObject | undefined = (
  path: string
) => (props: StyledInterpolationProps) => {
  const componentStyle = get(props.theme, `${componentStylesRoot}.${path}`)
  console.warn("CS: ", componentStyle)
  return componentStyle && styledCss(componentStyle)
}

const rxHandler: (p: StyledInterpolationProps) => CSSObject | undefined = ({
  rx,
}) => rx && styledCss(rx)

type VariantHandler = (
  key: string,
  op: ComposeOptions
) => (p: RosesStyleProps & StyledInterpolationProps) => CSSObject | undefined

const variantHandler: VariantHandler = (componentKey, ops) => ({
  theme,
  variant,
}) => {
  const selectedVariant = variant || ops.defaultVariant
  if (selectedVariant) {
    const base = `${themeVariantsRoot}.${componentKey}`
    const variantCss = get(theme, `${base}.${selectedVariant}`)
    console.warn("variant: ", variantCss)
    return variantCss && styledCss(variantCss)
  }
}

/**
 * Wrap a base styled component with:
 *   - a responsive, theme-aware style `css` prop
 *   - a `variant` prop for accessing styles set on the theme.variants key
 * @example:
 * ```tsx
 *  const RawBox = styled('div')({boxSizing: 'border-box'})
 *  const Widget = withStyleProps('Box', RawBox, {})
 * ```
 */
export const withStyleProps = (
  componentKey: string,
  Component: React.ComponentType<RosesStyleProps>,
  options: ComposeOptions = {} // not using this yet
) => {
  return styled(Component)<RosesStyleProps>(
    themedComponent(componentKey),
    variantHandler(componentKey, options),
    rxHandler
  )
}
// /** Compose in the default style + variants for the key defined in your theme
//  * @example ```
//  * const theme = {
//  *   styles: {
//  *     Button: {
//  *       color: "blue"
//  *     }
//  *   }
//  * }
//  *
//  * const Button = styled('button')(ButtonProps, themed("Button"))
//  * ```
//  */
// export const themed = key => props => props.theme.styles[key]
