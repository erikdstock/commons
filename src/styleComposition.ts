import styled from "@emotion/styled"
import ssCss, { CSSObject, SystemStyleObject } from "@styled-system/css"
import get from "lodash/get"
import React from "react"
import { RosesThemeObject } from "./Theme"
import { RosesSC, RosesStyledComponent, RosesStyleProps } from "./types"
import isPropValid from "@emotion/is-prop-valid"

/** The base style for Box and everything on up. */
export const boxStyle: CSSObject = {
  boxSizing: "border-box",
}

/** Clumsily force styledCss function to believe that there will be a theme present */
type StyledCssFn = (o: SystemStyleObject) => CSSObject
const styledCss = ssCss as StyledCssFn

const themeVariantsRoot: keyof RosesThemeObject = "variants"
const componentStylesRoot: keyof RosesThemeObject = "componentStyles"

interface StyledInterpolationProps extends RosesStyleProps {
  theme: RosesThemeObject
}

type InterpolationFn = (p: StyledInterpolationProps) => CSSObject | undefined

const relevantStyles: (
  path: string
) => (
  props: RosesStyleProps & { theme: RosesThemeObject }
) => CSSObject = themePath => ({ theme, variant }): CSSObject => {
  const { variants, ...styles } = get(theme, themePath, {})
  if (styles) {
    if (variant && variants && Object.keys(variants).length) {
      const selectedVariant = variants[variant]

      return Object.keys(selectedVariant).length
        ? styledCss({ ...styles, ...selectedVariant })
        : styledCss(styles)
    } else {
      return styledCss(styles)
    }
  } else return {}
}

const themeDiv = (name: string): RosesSC => {
  return styled("div", { shouldForwardProp: prop => isPropValid(prop) })<
    RosesStyleProps
  >(boxStyle, relevantStyles(`${componentStylesRoot}.${name}`), rxHandler)
}

const rxHandler: InterpolationFn = ({ rx }) => {
  return rx && styledCss(rx)
}

interface ThemedComponentOptions {
  name: string
  shouldForwardProp?: typeof isPropValid
  defaultVariant?: string
  component?: React.ComponentType<any> | keyof JSX.IntrinsicElements
}

export const themed = (
  options: ThemedComponentOptions | string
): RosesStyledComponent => {
  if (typeof options === "string") {
    const name = options as keyof JSX.IntrinsicAttributes
    return themeDiv(name)
  } else {
    const {
      component,
      shouldForwardProp = (prop: string) => prop !== "rx" && prop !== "variant",
      name,
    } = options

    return component
      ? styled(component as React.ComponentType<any>, { shouldForwardProp })<
          RosesStyleProps
        >(relevantStyles(`${componentStylesRoot}.${name}`), rxHandler)
      : themeDiv(`${componentStylesRoot}.${name}`)
  }
}

/** Reach into the theme, get a style object, apply it with styledCss */
const getThemeStyles: (path: string) => InterpolationFn = path => props => {
  const componentStyle = get(props.theme, path)
  return componentStyle && styledCss(componentStyle)
}

const getComponentStyles: (name: string) => InterpolationFn = (
  name: string
) => {
  return getThemeStyles(`${componentStylesRoot}.${name}`)
}

type VariantHandler = (key: string, defaultVariant?: string) => InterpolationFn

const variantHandler: VariantHandler = (componentKey, defaultVariant) => ({
  theme,
  variant,
}) => {
  const selectedVariant = variant || defaultVariant
  if (selectedVariant) {
    const base = `${themeVariantsRoot}.${componentKey}`
    const variantCss = get(theme, `${base}.${selectedVariant}`)
    return variantCss && styledCss(variantCss)
  }
}

const themedBoxDiv: (n: string, defaultVariant?: string) => RosesSC = (
  name,
  defaultVariant
) =>
  styled("div")<RosesStyleProps>(
    boxStyle,
    getComponentStyles(name),
    variantHandler(name, defaultVariant),
    rxHandler
  )

/**
 * Wrap a base styled component with:
 *   - a responsive, theme-aware style `rx` prop
 *   - a `variant` prop for accessing styles set on the theme.variants key
 * @example:
 *
 *  const RawBox = styled('div')({boxSizing: 'border-box'})
 *
 *  const Box = withStyleProps({name: 'Box', component: RawBox, defaultVariant: "hot"})
 *  const Widget = withStyleProps('Widget')
 *
 */
export const withStyleProps: (
  options: ThemedComponentOptions | string
) => RosesSC = options => {
  if (typeof options === "string") {
    const name = options as keyof JSX.IntrinsicElements
    return themedBoxDiv(name)
  } else {
    const component = options.component

    return component
      ? styled(component as React.ComponentType<any>, {
          shouldForwardProp: propName =>
            propName !== "rx" && propName !== "variant",
        })<RosesStyleProps>(
          getComponentStyles(options.name),
          variantHandler(options.name, options.defaultVariant),
          rxHandler
        )
      : themedBoxDiv(options.name, options.defaultVariant)
  }
}
