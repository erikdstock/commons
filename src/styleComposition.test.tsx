/** @jsx jsx */
import { jsx } from "@emotion/core"
import { withStyleProps, themed } from "./styleComposition"
import { RosesTheme } from "./Theme"
import renderer from "react-test-renderer"
import { baseTheme } from "./Theme/baseTheme"
import { matchers } from "jest-emotion"

const { colors: baseColors } = baseTheme as any
const theme = {
  ...baseTheme,
  colors: {
    ...baseColors,
    primary: "red",
    secondary: "green",
  },
}

expect.extend(matchers)

const render = (contents: React.ReactNode) =>
  renderer.create(<RosesTheme theme={theme}>{contents}</RosesTheme>)

describe("withStyleProps", () => {
  it("adds an rx prop that takes styled-system/css inputs", () => {
    const Widget = withStyleProps("div")
    const tree = render(<Widget rx={{ bg: "secondary" }} />)
    expect(tree.toJSON()).toMatchSnapshot()
    // expect(tree).toHaveStyleRule("background", "brown") //, { target: `${Widget}` })
  })
})

describe("themed()", () => {
  const localTheme = {
    ...theme,
    componentStyles: {
      Widget: {
        color: "primary",
        variants: {
          experimental: {
            color: "secondary",
          },
        },
      },
    },
  }

  it("adds component themes + variants from the theme", () => {
    const DivWidget = themed("Widget")
    const tree = renderer.create(
      <RosesTheme theme={localTheme}>
        <DivWidget />
      </RosesTheme>
    )
    expect(tree.toJSON()).toMatchSnapshot()

    const tree2 = renderer.create(
      <RosesTheme theme={theme}>
        <DivWidget variant="experimental" />
      </RosesTheme>
    )
    expect(tree2.toJSON()).toMatchSnapshot()
  })
})
