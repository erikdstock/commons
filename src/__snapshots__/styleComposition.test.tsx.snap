// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`themed() adds component themes + variants from the theme 1`] = `
.emotion-0 {
  box-sizing: border-box;
  color: red;
}

<div
  className="emotion-0"
/>
`;

exports[`themed() adds component themes + variants from the theme 2`] = `
.emotion-0 {
  box-sizing: border-box;
}

<div
  className="emotion-0"
/>
`;

exports[`withStyleProps adds an rx prop that takes styled-system/css inputs 1`] = `
.emotion-0 {
  box-sizing: border-box;
  background-color: green;
}

<div
  className="emotion-0"
  rx={
    Object {
      "bg": "secondary",
    }
  }
/>
`;
